### PR TITLE
add npy datatype

### PR DESCRIPTION
**Pull request by astrovsky01**
Adds numpy datatype for use in tools

### CODE REVIEW

**Comment by bernt-matthias** [comment]
Would be cool to have this versioning https://numpy.org/devdocs/reference/generated/numpy.lib.format.html#format-version-1-0

**Comment by astrovsky01** [comment]
@bernt-matthias as in via a sniffer or multiple datatypes representing version?

**Comment by bernt-matthias** [comment]
My preference would be metadata. Probably one would need a sniffer and set_meta function.

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -4538,3 +4538,53 @@ def display_peek(self, dataset: DatasetProtocol) -> str:
...
+    file_ext = "npy"
+
+    MetadataElement(
+        name="version",
+        default="",
+        param=DictParameter,
+        desc="Version string for the numpy file format",
+        readonly=True,
+        visible=True,
+        no_value={},
```
Comment (bernt-matthias) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
        default=0,
        param=MetadataParameter,
        desc="Version string for the numpy file format",
        readonly=True,
        visible=True,
        no_value=0,
```

File: lib/galaxy/datatypes/binary.py [hunk 2]
```diff
@@ -4538,3 +4538,53 @@ def display_peek(self, dataset: DatasetProtocol) -> str:
...
+        optional=True
+    )
+    def _numpy_version_string(self, filename):
+        magic_string = open(filename, "rb").read(8)
+        version_str = str(magic_string[6])+"."+str(magic_string[7])
+        return version_str
+    
+    def set_meta(self, dataset: DatasetProtocol, *, overwrite: TYPE_CHECKING = True, **kwd) -> None:
+        dataset.metadata.version_dict = self._numpy_version_string(dataset.get_file_name())
```
Comment (bernt-matthias) [inline lib/galaxy/datatypes/binary.py]: We should probably put this in a try-except block.

File: lib/galaxy/config/sample/datatypes_conf.xml.sample [hunk 3]
```diff
@@ -220,6 +220,7 @@
     <datatype extension="ktab" type="galaxy.datatypes.binary:Binary" subclass="true" description="A table of canonical k‑mers and their counts for the fastk toolkit." display_in_upload="true" description_url="https://github.com/thegenemyers/FASTK?tab=readme-ov-file#file-encodings"/>
     <datatype extension="hist" type="galaxy.datatypes.binary:Binary" subclass="true" description="A binary histogram file of kmers and frequencies for the fastk toolkit." display_in_upload="true" description_url="https://github.com/thegenemyers/FASTK?tab=readme-ov-file#file-encodings"/>
     <datatype extension="prof" type="galaxy.datatypes.binary:Binary" subclass="true" description="Read profile file for the fastk toolkit." display_in_upload="true" description_url="https://github.com/thegenemyers/FASTK?tab=readme-ov-file#file-encodings"/>
+    <datatype extension="npy" type="galaxy.datatypes.binary:Binary" subclass="true" description="Standard format for saving numpy arrays" display_in_upload="true" description_url="https://numpy.org/devdocs/reference/generated/numpy.lib.format.html"/>
```
Comment (mvdbeek) [inline lib/galaxy/config/sample/datatypes_conf.xml.sample]:
```suggestion
    <datatype extension="npy" type="galaxy.datatypes.binary:Numpy" description="Standard format for saving numpy arrays" display_in_upload="true" description_url="https://numpy.org/devdocs/reference/generated/numpy.lib.format.html"/>
```

File: lib/galaxy/datatypes/binary.py [hunk 4]
```diff
@@ -4538,3 +4538,58 @@ def display_peek(self, dataset: DatasetProtocol) -> str:
...
+        desc="Version string for the numpy file format",
+        readonly=True,
+        visible=True,
+        no_value=0,
+        optional=True,
+    )
+
+    def _numpy_version_string(self, filename):
+        magic_string = open(filename, "rb").read(8)
+        version_str = str(magic_string[6]) + "." + str(magic_string[7])
```
Comment (VascoSch92) [inline lib/galaxy/datatypes/binary.py]: You can also a f-string here. It is much easier to read.

File: lib/galaxy/datatypes/binary.py [hunk 5]
```diff
@@ -4538,3 +4538,58 @@ def display_peek(self, dataset: DatasetProtocol) -> str:
...
+            log.warning("%s, set_meta Exception: %s", self, e)
+
+    def sniff_prefix(self, file_prefix: FilePrefix) -> bool:
+        # The first 6 bytes of any numpy file is '\x93NUMPY', with following bytes for version
+        # number of file formats, and info about header data. The rest of the file contains binary data.
+        return file_prefix.startswith_bytes(b"\x93NUMPY")
+
+    def set_peek(self, dataset: DatasetProtocol, **kwd) -> None:
+        if not dataset.dataset.purged:
+            dataset.peek = "Binary numpy file version %s" % dataset.metadata.version_str
```
Comment (VascoSch92) [inline lib/galaxy/datatypes/binary.py]: Also here f-string

File: lib/galaxy/datatypes/binary.py (hunk 4 above)
Comment (martenson) [reply lib/galaxy/datatypes/binary.py]:
```suggestion
        version_str = f"{magic_string[6]}.{magic_string[7]}"
```

File: lib/galaxy/datatypes/binary.py (hunk 5 above)
Comment (martenson) [reply lib/galaxy/datatypes/binary.py]:
```suggestion
            dataset.peek = f"Binary numpy file version {dataset.metadata.version_str}"
```

**Comment by astrovsky01** [comment]
Thank you @martenson and @VascoSch92!

**Review by jdavcs — APPROVED** [review]

**Pull request merged**